### ISSUE TITLE
Healthcheck Timeouts - Destination Container Only

### ISSUE THREAD

**Issue by zer0stars** [linkerd/linkerd2#8760]
### What is the issue?

```
7m45s       Warning   Unhealthy   pod/linkerd-destination-676b6657b-426bs   Readiness probe failed: Get "http://10.0.28.9:9997/ready": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
56m         Warning   Unhealthy   pod/linkerd-destination-676b6657b-426bs   Readiness probe failed: Get "http://10.0.28.9:9990/ready": read tcp 10.0.21.44:60324->10.0.28.9:9990: read: connection reset by peer
58m         Warning   Unhealthy   pod/linkerd-destination-676b6657b-6v89v   Readiness probe failed: Get "http://10.0.50.176:9997/ready": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

### How can it be reproduced?

Running latest stable prod release on 1.22.9 Kubernetes nodes.

### Logs, error output, etc

 pod/linkerd-destination-676b6657b-6v89v   Readiness probe failed: Get "http://10.0.50.176:9997/ready

### output of `linkerd check -o short`

```
Linkerd core checks
===================

kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ control plane pods are ready
‼ cluster networks can be verified
    the following nodes do not expose a podCIDR:
	ip-10-0-21-44.us-east-2.compute.internal
	ip-10-0-35-11.us-east-2.compute.internal
	ip-10-0-45-81.us-east-2.compute.internal
	ip-10-0-93-1.us-east-2.compute.internal
    see https://linkerd.io/2.11/checks/#l5d-cluster-networks-verified for hints

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ proxy-init container runs as root user if docker container runtime is used

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
‼ issuer cert is valid for at least 60 days
    issuer certificate will expire on 2022-06-30T02:40:10Z
    see https://linkerd.io/2.11/checks/#l5d-identity-issuer-cert-not-expiring-soon for hints
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days
√ policy-validator webhook has valid cert
√ policy-validator cert is valid for at least 60 days

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ can retrieve the control plane version
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-control-plane-proxy
---------------------------
√ control plane proxies are healthy
√ control plane proxies are up-to-date
√ control plane proxies and cli versions match

linkerd-ha-checks
-----------------
√ pod injection disabled on kube-system

Linkerd extensions checks
=========================

linkerd-viz
-----------
√ linkerd-viz Namespace exists
√ linkerd-viz ClusterRoles exist
√ linkerd-viz ClusterRoleBindings exist
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ tap API service is running
√ linkerd-viz pods are injected
√ viz extension pods are running
√ viz extension proxies are healthy
√ viz extension proxies are up-to-date
√ viz extension proxies and cli versions match
‼ prometheus is installed and configured correctly
    missing ClusterRoles: linkerd-linkerd-viz-prometheus
    see https://linkerd.io/2.11/checks/#l5d-viz-prometheus for hints
√ can initialize the client
√ viz extension self-check

Status check results are √
```

### Environment

EKS 1.22.9 - Bottlerocket

### Would you like to work on fixing this bug?

no

**Comment by adleong** [comment]
Hi @zer0stars!  The next steps for investigating this would be to take a look at the logs of the policy-controller container in the destination pod as well as the linkerd-proxy container in that pod.  Hopefully those logs will provide some clue as to why the health checks are failing/timing out.

**Comment by zer0stars** [comment]
[linkerd_proxy_logs.log](https://github.com/linkerd/linkerd2/files/9043041/linkerd_proxy_logs.log)
Attaching logs for the proxy.  I did not find anything in the logs of the destination controller(request doesn't reach there). One thing that stands out is number of connection resets.

**Comment by zer0stars** [comment]
Adding some additional info that I found a little odd. 
1. The load between two linkers-destination pods is not even. For some reason one of the pods is receiving majority of the traffic.
<img width="1645" alt="image" src="https://user-images.githubusercontent.com/74260741/177600899-e8548857-94d1-4603-b3ab-aaffc281360c.png">
2. Why is there so many sockets open?3. 
3. Attaching the conntrack table. [conntrack.txt](https://github.com/linkerd/linkerd2/files/9056657/conntrack.txt)

**Comment by zer0stars** [comment]
Adding sysctl config here from the node:

net.mptcp.enabled = 1
net.netfilter.nf_conntrack_acct = 0
net.netfilter.nf_conntrack_buckets = 65536
net.netfilter.nf_conntrack_checksum = 1
net.netfilter.nf_conntrack_count = 2572
net.netfilter.nf_conntrack_dccp_loose = 1
net.netfilter.nf_conntrack_dccp_timeout_closereq = 64
net.netfilter.nf_conntrack_dccp_timeout_closing = 64
net.netfilter.nf_conntrack_dccp_timeout_open = 43200
net.netfilter.nf_conntrack_dccp_timeout_partopen = 480
net.netfilter.nf_conntrack_dccp_timeout_request = 240
net.netfilter.nf_conntrack_dccp_timeout_respond = 480
net.netfilter.nf_conntrack_dccp_timeout_timewait = 240
net.netfilter.nf_conntrack_events = 1
net.netfilter.nf_conntrack_expect_max = 1024
net.netfilter.nf_conntrack_frag6_high_thresh = 4194304
net.netfilter.nf_conntrack_frag6_low_thresh = 3145728
net.netfilter.nf_conntrack_frag6_timeout = 60
net.netfilter.nf_conntrack_generic_timeout = 120
net.netfilter.nf_conntrack_gre_timeout = 30
net.netfilter.nf_conntrack_gre_timeout_stream = 180
net.netfilter.nf_conntrack_helper = 0
net.netfilter.nf_conntrack_icmp_timeout = 30
net.netfilter.nf_conntrack_icmpv6_timeout = 30
net.netfilter.nf_conntrack_log_invalid = 0
net.netfilter.nf_conntrack_max = 131072
net.netfilter.nf_conntrack_sctp_timeout_closed = 10
net.netfilter.nf_conntrack_sctp_timeout_cookie_echoed = 3
net.netfilter.nf_conntrack_sctp_timeout_cookie_wait = 3
net.netfilter.nf_conntrack_sctp_timeout_established = 432000
net.netfilter.nf_conntrack_sctp_timeout_heartbeat_acked = 210
net.netfilter.nf_conntrack_sctp_timeout_heartbeat_sent = 30
net.netfilter.nf_conntrack_sctp_timeout_shutdown_ack_sent = 3
net.netfilter.nf_conntrack_sctp_timeout_shutdown_recd = 0
net.netfilter.nf_conntrack_sctp_timeout_shutdown_sent = 0
net.netfilter.nf_conntrack_tcp_be_liberal = 0
net.netfilter.nf_conntrack_tcp_loose = 1
net.netfilter.nf_conntrack_tcp_max_retrans = 3
net.netfilter.nf_conntrack_tcp_timeout_close = 10
net.netfilter.nf_conntrack_tcp_timeout_close_wait = 3600
net.netfilter.nf_conntrack_tcp_timeout_established = 86400
net.netfilter.nf_conntrack_tcp_timeout_fin_wait = 120
net.netfilter.nf_conntrack_tcp_timeout_last_ack = 30
net.netfilter.nf_conntrack_tcp_timeout_max_retrans = 300
net.netfilter.nf_conntrack_tcp_timeout_syn_recv = 60
net.netfilter.nf_conntrack_tcp_timeout_syn_sent = 120
net.netfilter.nf_conntrack_tcp_timeout_time_wait = 120
net.netfilter.nf_conntrack_tcp_timeout_unacknowledged = 300
net.netfilter.nf_conntrack_timestamp = 0
net.netfilter.nf_conntrack_udp_timeout = 30
net.netfilter.nf_conntrack_udp_timeout_stream = 120

**Comment by zer0stars** [comment]
Tried to change net.ipv4.tcp_tw_reuse=2 to net.ipv4.tcp_tw_reuse=1 to see if that would make an improvement, but I am still facing this issue. 

This is on k8s 1.22, worker node 1.22.9.

**Comment by zer0stars** [comment]
Bumping this, we are still experiencing errors in destination/identity pods. After watching this issue for several weeks, it actually has started occurring in other high connection count listeners. The error is still the same "connection closed before message completed".

**Comment by jaygridley** [comment]
I think the fix in https://github.com/linkerd/linkerd2/releases/tag/edge-22.8.3 could help with this

**Comment by irizzant** [comment]
Is there a planned stable relase with this fix? Currently I'm hitting this issue on stable 2.12.0 and it's causing a flood of failed probes events

**Comment by kleimkuhler** [comment]
Yes this fix will be included in `stable-2.12.1` and should be released soon.
